database/tbcd/level/level.go

```diff
@@ -241,137 +266,372 @@ func keyToHeightHash(key []byte) (uint64, []byte) {
+	// Reason about needing to check fork flag. For now keep it here to
+	// distinguish between certain fork and maybe fork paths.
+	var (
+		it       tbcd.InsertType
+		cbh, lbh *tbcd.BlockHeader
+	)
+	if fork {
+		// Insert last height into block headers if the new cumulative
+		// difficulty exceeds the prior difficulty.
+		switch cdiff.Cmp(&bestBH.Difficulty) {
+		case -1, 0:
+			// Extend fork, fork did not overcome difficulty
+			it = tbcd.ITForkExtend
+
+			// XXX should we return old best block header here?
+			// That way the caller can do best vs previous best diff.
+			log.Debugf("(%v) : %v <= %v", height, cdiff, bestBH.Difficulty)
+			cbh = bestBH
+			var header [80]byte
+			lbh = &tbcd.BlockHeader{
+				Hash:       bhash[:],
+				Height:     height,
+				Header:     header[:],
+				Difficulty: *cdiff,
+			}
+
+		case 1:
+			log.Debugf("(%v) 1: %v > %v", height, cdiff, bestBH.Difficulty)
+			// log.Infof("%v", spew.Sdump(bestBH.Hash[:]))
+			// log.Infof("%v", spew.Sdump(firstHash))
+			// pick the right return value based on ancestor
+			bhsBatch.Put([]byte(bhsCanonicalTipKey), lastRecord)
+			it = tbcd.ITChainFork
+
+			// XXX duplicate below
+			var header [80]byte
+			copy(header[:], bhs[len(bhs)-1][:])
+			cbh = &tbcd.BlockHeader{
+				Hash:       bhash[:],
+				Height:     height,
+				Header:     header[:],
+				Difficulty: *cdiff,
+			}
+			lbh = cbh
+
+		default:
+			panic("bug: impossible cmp value")
+		}
+	} else {
+		// Extend current best tip
+		bhsBatch.Put([]byte(bhsCanonicalTipKey), lastRecord)
+		it = tbcd.ITChainExtend
+
+		// XXX duplicate above
+		var header [80]byte
+		copy(header[:], bhs[len(bhs)-1][:])
+		cbh = &tbcd.BlockHeader{
+			Hash:       bhash[:],
+			Height:     height,
+			Header:     header[:],
+			Difficulty: *cdiff,
+		}
+		lbh = cbh
+	}
 
 	// Write height hash batch
 	err = hhTx.Write(hhBatch, nil)
 	if err != nil {
-		return fmt.Errorf("height hash batch: %w", err)
+		return tbcd.ITInvalid, nil, nil,
+			fmt.Errorf("height hash batch: %w", err)
 	}
 
 	// Write missing blocks batch
 	err = bmTx.Write(bmBatch, nil)
 	if err != nil {
-		return fmt.Errorf("blocks missing batch: %w", err)
+		return tbcd.ITInvalid, nil, nil,
+			fmt.Errorf("blocks missing batch: %w", err)
 	}
 
 	// Write block headers batch
 	err = bhsTx.Write(bhsBatch, nil)
 	if err != nil {
-		return fmt.Errorf("block headers insert: %w", err)
+		return tbcd.ITInvalid, nil, nil,
+			fmt.Errorf("block headers insert: %w", err)
 	}
 
 	// height hash commit
 	err = hhCommit()
 	if err != nil {
-		return fmt.Errorf("height hash commit: %w", err)
+		return tbcd.ITInvalid, nil, nil,
+			fmt.Errorf("height hash commit: %w", err)
 	}
 
 	// blocks missing commit
 	err = bmCommit()
 	if err != nil {
-		return fmt.Errorf("blocks missing commit: %w", err)
+		return tbcd.ITInvalid, nil, nil,
+			fmt.Errorf("blocks missing commit: %w", err)
 	}
 
 	// block headers commit
 	err = bhsCommit()
 	if err != nil {
```

Comment:
I think this error assignment could be nested:
```suggestion
	if err = bhsCommit(); err != nil {
```